starfish/codebook.py

```diff
@@ -0,0 +1,545 @@
+
+        Examples
+        --------
+        >>> from starfish.constants import Indices
+        >>> from starfish.codebook import Codebook
+
+        >>> codebook = [
+        >>>     {
+        >>>         Codebook.Constants.CODEWORD.value: [
+        >>>             {Indices.HYB.value: 0, Indices.CH.value: 3, Codebook.Constants.VALUE.value: 1},
+        >>>             {Indices.HYB.value: 1, Indices.CH.value: 3, Codebook.Constants.VALUE.value: 1},
+        >>>         ],
+        >>>         Codebook.Constants.GENE.value: "ACTB_human"
+        >>>     },
+        >>>     {
+        >>>         Codebook.Constants.CODEWORD.value: [
+        >>>             {Indices.HYB.value: 0, Indices.CH.value: 3, Codebook.Constants.VALUE.value: 1},
+        >>>             {Indices.HYB.value: 1, Indices.CH.value: 1, Codebook.Constants.VALUE.value: 1},
+        >>>         ],
+        >>>         Codebook.Constants.GENE.value: "ACTB_mouse"
+        >>>     },
+        >>> ]
+
+        >>> Codebook.from_json(codebook)
+       <xarray.Codebook (gene_name: 2, c: 4, h: 2)>
+        array([[[0, 0],
+                [0, 0],
+                [0, 0],
+                [1, 1]],
+
+               [[0, 0],
+                [0, 1],
+                [0, 0],
+                [1, 0]]], dtype=uint8)
+        Coordinates:
+          * gene_name  (gene_name) object 'ACTB_human' 'ACTB_mouse'
+          * c          (c) int64 0 1 2 3
+          * h          (h) int64 0 1
+
+        Returns
+        -------
+        Codebook :
+            Codebook with shape (genes, channels, hybridization_rounds)
+
+        """
+        with open(json_codebook, 'r') as f:
+            code_array = json.load(f)
+        return cls.from_code_array(code_array, n_hyb, n_ch)
+
+    def to_json(self, filename: str) -> None:
+        """save a codebook to json
+
+        Notes
+        -----
+        This enforces the following typing of codebooks:
+        ch, hyb: int
+        value: float
+        gene: str
+
+        Parameters
+        ----------
+        filename : str
+            filename
+
+        """
+        code_array = []
+        for gene in self[self.Constants.GENE.value]:
+            codeword = []
+            for ch in self[Indices.CH.value]:
+                for hyb in self[Indices.HYB.value]:
+                    if self.loc[gene, ch, hyb]:
+                        codeword.append(
+                            {
+                                Indices.CH.value: int(ch),
+                                Indices.HYB.value: int(hyb),
+                                self.Constants.VALUE.value: float(self.loc[gene, ch, hyb])
+                            })
+            code_array.append({
+                self.Constants.CODEWORD.value: codeword,
+                self.Constants.GENE.value: str(gene.values)
+            })
+
+        with open(filename, 'w') as f:
+            json.dump(code_array, f)
+
+    def decode_euclidean(self, intensities: IntensityTable) -> IntensityTable:
```

Comment:
Intuitively, I think it's weird to decode from and then write to the same type, but I get why that is.  Don't have a concrete proposal/alternative though.

Reply:
Agree. Open to ideas. I thought about returning the gene vector, but then we're stuck with another object, and they fit so nicely into the feature metadata.